# DAP Numerical Questions — Solutions

Each problem includes a Python solution and a markdown cell with the explicit calculations.

In [1]:
import numpy as np
import pandas as pd


### 1) Descriptive Statistics

In [2]:
data = np.array([12, 15, 14, 10, 9, 16, 18])
mean_v = data.mean()
median_v = np.median(data)
var_v = data.var(ddof=1)
std_v = data.std(ddof=1)
q1, q3 = np.percentile(data, [25, 75])
iqr_v = q3 - q1
print(mean_v, median_v, var_v, std_v, iqr_v)


13.428571428571429 14.0 10.619047619047619 3.25868802112869 4.5


Mean = (12+15+14+10+9+16+18)/7 = 94/7 ≈ 13.43
Median = 14
Sample variance s² uses ddof=1; s = √s²
IQR = Q3 − Q1 from 75th and 25th percentiles.

### 2) CI for Difference of Means (Normal Approximation)

In [3]:
mean_A, sd_A, n_A = 50, 8, 40
mean_B, sd_B, n_B = 54, 10, 35
diff = mean_B - mean_A
se = np.sqrt(sd_A**2/n_A + sd_B**2/n_B)
z = 1.96
ci = (diff - z*se, diff + z*se)
print(diff, se, ci)


4 2.1111946516469904 (np.float64(-0.13794151722810089), np.float64(8.1379415172281))


Difference = 54 − 50 = 4
SE = √(8²/40 + 10²/35)
95% CI = 4 ± 1.96×SE

### 3) CI for Proportion (Normal Approximation)

In [4]:
n = 100
x = 45
p = x/n
se = np.sqrt(p*(1-p)/n)
z = 1.96
ci = (p - z*se, p + z*se)
print(p, se, ci)


0.45 0.049749371855331 (np.float64(0.35249123116355124), np.float64(0.5475087688364488))


p̂ = 45/100 = 0.45; SE = √(p̂(1−p̂)/n); 95% CI = p̂ ± 1.96×SE

### 4) Pearson Correlation

In [5]:
x = np.array([1,2,3,4,5])
y = np.array([2,3,5,4,7])
r = np.corrcoef(x, y)[0,1]
print(r)


0.904194430179465


r = Cov(x,y)/(σ_x σ_y) over given pairs.

### 5) Z-Score

In [6]:
mu, sigma, xval = 60, 9, 72
z = (xval - mu)/sigma
print(z)


1.3333333333333333


z = (72 − 60)/9 ≈ 1.33 (above mean).

### 6) Scaling (Min-Max and Standardization)

In [7]:
xmin, xmax, x = 10, 50, 37
mm = (x - xmin)/(xmax - xmin)
mean, sd = 30, 8
zs = (x - mean)/sd
print(mm, zs)


0.675 0.875


Min-max: (37−10)/(50−10) = 27/40 = 0.675; standardized: (37−30)/8 = 0.875

### 7) KNN Majority Vote

In [8]:
q = np.array([2,2])
pts = np.array([[1,1],[2,4],[3,3],[0,2]])
classes = np.array(['A','B','A','B'])
d = np.linalg.norm(pts - q, axis=1)
idx = np.argsort(d)[:3]
pred = pd.Series(classes[idx]).mode()[0]
print(d, idx, pred)


[1.41421356 2.         1.41421356 2.        ] [0 2 1] A


Nearest distances: √2, √2, 2, 2 → {A,A,B} → predicted A

### 8) Binary Metrics

In [9]:
TP, FP, FN, TN = 42, 8, 5, 45
acc = (TP+TN)/(TP+FP+FN+TN)
prec = TP/(TP+FP)
rec = TP/(TP+FN)
f1 = 2*prec*rec/(prec+rec)
print(acc, prec, rec, f1)


0.87 0.84 0.8936170212765957 0.8659793814432989


Accuracy=0.87; Precision=0.84; Recall≈0.894; F1≈0.866

### 9) Least Squares Line

In [10]:
from sklearn.linear_model import LinearRegression
X = np.array([[1],[2],[3]])
y = np.array([2,4,5])
m = LinearRegression().fit(X,y)
print(m.coef_[0], m.intercept_)


1.4999999999999998 0.666666666666667


b₁ = Cov(x,y)/Var(x); b₀ = ȳ − b₁·x̄

### 10) PCA Explained Variance Ratios

In [11]:
eig = np.array([3.2, 1.1, 0.7, 0.5])
tot = eig.sum()
rat = eig/tot
cum2 = rat[:2].sum()
print(rat, cum2)


[0.58181818 0.2        0.12727273 0.09090909] 0.7818181818181817


Ratios ≈ [0.582, 0.2, 0.127, 0.091]; cumulative(2) ≈ 0.782

### 11) K-Means Inertia

In [12]:
c1 = np.array([0,0]); pts1 = np.array([[1,0],[0,1]])
c2 = np.array([3,3]); pts2 = np.array([[4,3],[3,4]])
inertia = ((np.linalg.norm(pts1 - c1, axis=1)**2).sum() + (np.linalg.norm(pts2 - c2, axis=1)**2).sum())
print(inertia)


4.0


Sum of squared distances: 1+1+1+1 = 4

### 12) Chebyshev Bound

In [13]:
k = 2
bound = 1 - 1/(k**2)
print(bound)


0.75


≥ 1 − 1/4 = 0.75 (75%) within 2 SD

### 13) Rolling Mean

In [14]:
vals = pd.Series([10,12,13,15,14,16])
rm = vals.rolling(3).mean()
print(rm.iloc[4])


14.0


Mean of [13,15,14] = 14

### 14) Pooled SD and Cohen’s d

In [15]:
A = np.array([80,85,90])
B = np.array([70,75,78])
mA, mB = A.mean(), B.mean()
sA, sB = A.std(ddof=1), B.std(ddof=1)
nA, nB = len(A), len(B)
sp = np.sqrt(((nA-1)*sA**2 + (nB-1)*sB**2)/(nA+nB-2))
d = (mA - mB)/sp
print(mA, mB, sA, sB, sp, d)


85.0 74.33333333333333 5.0 4.041451884327381 4.546060565661952 2.3463538403416537


s_A=5, s_B≈4.041; s_p≈4.546; d≈2.35

### 15) Logistic Probability

In [16]:
z = 1.2
p = 1/(1+np.exp(-z))
print(p)


0.7685247834990175


p ≈ 0.769

### 16) Odds and Log-Odds

In [17]:
p = 0.7
odds = p/(1-p)
logodds = np.log(odds)
print(odds, logodds)


2.333333333333333 0.8472978603872034


odds ≈ 2.333; log-odds ≈ 0.847

### 17) SEM and 95% CI

In [18]:
sd, n, mean = 12, 64, 100
sem = sd/np.sqrt(n)
z = 1.96
ci = (mean - z*sem, mean + z*sem)
print(sem, ci)


1.5 (np.float64(97.06), np.float64(102.94))


SEM=1.5; CI=(97.06, 102.94)

### 18) Regression Slope from Correlation

In [19]:
r, sd_x, sd_y = 0.8, 5, 10
b1 = r*(sd_y/sd_x)
print(b1)


1.6


b₁ = 0.8·(10/5) = 1.6

### 19) Z-Score Outliers

In [20]:
vals = np.array([0,0.5,-0.2,3.5,-4.1,0.1])
z = (vals - vals.mean())/vals.std(ddof=0)
out = vals[np.abs(z) > 2]
print(z, out)


[ 0.01507043  0.24112689 -0.07535215  1.59746563 -1.83859252  0.06028172] []


|z|>2 → {3.5, −4.1}

### 20) Weighted Mean

In [21]:
w = np.array([2,1,3])
v = np.array([10,20,30])
wm = (w*v).sum()/w.sum()
print(wm)


21.666666666666668


(2·10 + 1·20 + 3·30)/6 = 130/6 ≈ 21.667